### ISSUE TITLE
dask.base's normalize_array hashes the same data to different values

### ISSUE THREAD

**Issue by dlovell** [dask/dask#2498]
https://github.com/dask/dask/blob/0.15.0/dask/base.py#L400

Is there a reason to both hash the ravel'ed data and use strides?  Reading some other issues suggests strides was introduce for memmaps but I think this borks the case where its just a plain np.array.  Also, passing an `order` kwarg to `ravel` makes the same data come out in different ways.

It looks like a recent change introduces an `order` kwarg to the `ravel` method call
Is it necessary for something?  

```python
import pprint

import dask.base
import dask.hashing
import numpy as np


def f(x):
    data = dask.hashing.hash_buffer_hex(x.ravel().view('i1'))
    return (data, x.dtype, x.shape)


data = [[0, 1], [2, 3]]
orders = ('C', 'F')
# arrs have the same data
arrs = [np.array(data, order=order) for order in orders]
# but tokenize differently
pprint.pprint([dask.base.tokenize(arr) for arr in arrs])
# the cause of the difference is passing `order` to ravel and using strides
pprint.pprint([dask.base.normalize_token(arr) for arr in arrs])
# what if we omit the `order` arg to ravel and the use of strides?
pprint.pprint([f(arr) for arr in arrs])
pprint.pprint([np.__version__, dask.__version__])
```
will print
```
['77acec63d80d37fcc8d42d0b4ee467bb', '85fb7b31b123920d0609d087c3d47300']
[('f5c356611bc74f63a6d8068bdb3b807ce19691fb', dtype('int64'), (2, 2), (16, 8)),
 ('f2837ee9691f05f519de044b5aae40589bec6155', dtype('int64'), (2, 2), (8, 16))]
[('f5c356611bc74f63a6d8068bdb3b807ce19691fb', dtype('int64'), (2, 2)),
 ('f5c356611bc74f63a6d8068bdb3b807ce19691fb', dtype('int64'), (2, 2))]
['1.11.3', '0.15.0+16.g24afbd5']
```

**Comment by mrocklin** [comment]
Is this behavior negatively affecting you in some way?  If so then are you able to describe how?

Some people would describe these two arrays as different because they have different strides.  Historically we have tended to err on the side of being pretty conservative when it comes to hashing conflicts.

**Comment by tsdlovell** [comment]
I think this is the crux of it:
```
arr = np.array([[1, 2], [3, 4]], order='f')
pprint.pprint((arr.strides, arr.copy().strides))
```
prints
```
((8, 16), (16, 8))
```

I think its fair to say that these are in fact different, though I would call it a bit surprising.

Perhaps there could be a switch to allow you to say you are "only numerically concerned" about equivalence?

**Comment by mrocklin** [comment]
You might be able to redefine and reregister the `normalize_token` function for numpy arrays to implement your own custom behavior.  See dask/base.py

**Comment by tsdlovell** [comment]
Yes, that does work.  I was suggesting the switch so that its more evident to other users that this might be an issue and so that its a part of dask (and gets whatever changes/updates are appropriate as other pieces change).

**Comment by mrocklin** [comment]
Adding a switch is convenient, but also comes at a cost.  There are hundreds or thousands of small choices like this.  Adding little switches everywhere tends to increase development inertia.  I'm more inclined to leave configuration at the more general `normalize_token` level just for code maintainability reasons.

**Comment by brianvanleeuwen** [comment]
@mrocklin , `np.ndarray.copy()` defaults to order=C.  If I assign `x=pd.DataFrame([[0,1],[2,3]]).values`, then  `tokenize(x)` is not equal to `tokenize(x.copy())` because of the order change.  Is there some reason we would want copies of arrays to have distinct tokens based on how they are ordered in memory?

**Comment by brianvanleeuwen** [comment]
There may be a minor bug on https://github.com/dask/dask/blob/master/dask/base.py#L399 due to the order=C default of np.ndarray.copy().  The same array could tokenize differently depending whether there is a BufferError, because `x.ravel(order='K').view('i1'))` could differ from `x.copy().ravel(order='K').view('i1'))`.  It might be safer to use `x.copy(order='K')`.

**Comment by mrocklin** [comment]
Adding `order='K'` seems like a decent approach as long as it produces a contiguous memoryview.

I'm curious, how is this affecting you all?  Is there an application where this is causing bad behavior?  I haven't come across a situation where hashing different variables with the same content to different keys produced incorrect results.  The only issue here that I see is that we sometimes lose out on a potential deduplication.  To me this is a minor inconvenience at worst.  I may not fully appreciate the situation though.  Can either of you help me to understand the relevance here?

**Comment by pitrou** [comment]
Just because some arrays contain the same data means they will behave the same. Trivially, imagine you execute the function `lambda a: a.strides`. Then this function will return different results for arrays having the same data (the same argument holds for the `dtype`).

**Comment by mrocklin** [comment]
Would adding the `order='K'` keyword to copy as suggested by @brianvanleeuwen resolve your issue?

**Comment by tsdlovell** [comment]
I think we would still need to re-register: there are enough calls to `copy` around, some we may not be able to control, that its "safer"/saner to just re-register `normalize_token`

**Comment by adamklein** [comment]
Ok, so we are going the route of reregistering `ndarray` for ourselves, but it's a bit subtle, because if we want to invoke 

        @normalize_token.register(np.ndarray)
        def layout_insensitive_normalize_array(x):

Then our registered function gets replaced when later another non-`ndarray` `numpy` class requests to be tokenized, and so the lazily registered `@normalize_token.register_lazy("numpy")` then gets invoked, which overwrites our earlier eager registration for `ndarray`. Basically we have to rewrite the whole lazy registration of numpy instead of just one hook. 

I think lazily loaded registrations should not overwrite eagerly loaded ones.
